[WIP] ID Translation

benbjohnson opened the pull request:
## Overview

This pull request adds id-to-key translation to make it easier for users
to provide non-integer identifiers for rows & columns.

_NOTE: This is a work-in-progress. Do not merge._

### TODO

- [x] Translator
- [x] TranslateStore
- [ ] Executor integration
- [ ] `main` integration


## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [ ] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [ ] I have resolved any merge conflicts.
- [x] I have included tests that cover my changes.
- [ ] All new and existing tests pass.

## Code review checklist

This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

benbjohnson commented:
@travisturner I refactored out the `Translator` and merged it into `Executor`. It seems to fit a lot better there. However, we need some kind of flag regarding whether to translate a query or not. Are we adding a flag per-frame or per-index to specify if string keys should be used?

jaffee commented:
For columns it must be per index. For rows, I think it should be per-frame.

benbjohnson commented:
@jaffee 👍 That makes sense. Any suggestions for what that flag should be called? Maybe just `Keys` or `UseKeys`?

jaffee commented:
I like "Keys" I think.

We were talking about this yesterday, and I think the consensus is that in PQL, if you are using keys, the arguments always must be quoted, and if you are providing your own sequential integers then the arguments MUST be unquoted. This will allow for the case of using integers as keys, but mapping them to sequential integers internally.

What I'm wondering now is if we can support both using keys and optionally not using them in the same frame/index. It might be nice for performance purposes if a client could decide to cache the whole mapping locally and use the integer API instead of the Key API for performance purposes. To be clear, I don't think we should pursue this initially as it adds a lot of complexity and corner cases, but I just want to keep it in the back of our heads as we design this API.

benbjohnson commented:
I added the schema support to `Index` & `Field`, fixed up the rebase, integrated the HTTP async replication, added CLI/config support, and did integration testing. There are still a few issues with some of the call types and how they interact with the bits/rows but several calls like `SetBit()` & `Bitmap()` are working.

Can this PR get merged in so I can do smaller PRs against `develop` from now on? :)

jaffee commented:
I'll look at it asap.

benbjohnson commented:
Thanks @jaffee

jaffee commented on server.go:
```diff
@@ -241,6 +251,14 @@ func NewServer(opts ...ServerOption) (*Server, error) {
+	s.TranslateFile = NewTranslateFile()
+	s.TranslateFile.Path = filepath.Join(path, "keys")
+	s.TranslateFile.PrimaryTranslateStore = s.primaryTranslateStore
+	if err := s.TranslateFile.Open(); err != nil {
```
I feel like this belongs in `Server.Open` rather than here in `NewServer`. Probably OK either way though.
